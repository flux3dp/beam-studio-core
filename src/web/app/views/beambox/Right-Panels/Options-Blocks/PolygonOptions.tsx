import React from 'react';

import i18n from 'helpers/i18n';
import ObjectPanelItem from 'app/views/beambox/Right-Panels/ObjectPanelItem';
import OptionPanelIcons from 'app/icons/option-panel/OptionPanelIcons';
import UnitInput from 'app/widgets/Unit-Input-v2';
import { useIsMobile } from 'helpers/system-helper';

import styles from './PolygonOptions.module.scss';

const LANG = i18n.lang.beambox.right_panel.object_panel.option_panel;

interface Props {
  elem: Element;
  polygonSides: number;
}

function PolygonOptions({ elem, polygonSides }: Props): JSX.Element {
  const [sides, setSides] = React.useState(polygonSides || 5);
  const isMobile = useIsMobile();
  React.useEffect(() => {
    if (polygonSides) setSides(polygonSides);
  }, [polygonSides]);

  const handleSideChanage = (val) => {
    if (val === sides) return;
    if (val > sides) {
      for (let i = sides; i < val; i += 1) window.polygonAddSides?.();
    } else {
      for (let i = val; i < sides; i += 1) window.polygonDecreaseSides?.();
    }
    setSides(+$(elem).attr('sides'));
  };

  const renderSides = () =>
    isMobile ? (
      <ObjectPanelItem.Number
        id="polygon-sides"
        value={sides}
        min={3}
        updateValue={handleSideChanage}
        label={LANG.sides}
        unit=""
        decimal={0}
      />
    ) : (
      <div className={styles['polygon-sides']} key="polygon-sides">
        <div className={styles.label} title={LANG.sides}>
          <OptionPanelIcons.PolygonSide />
        </div>
        <UnitInput
          min={3}
          className={{ 'option-input': true }}
          defaultValue={sides}
          decimal={0}
          getValue={(val) => handleSideChanage(val)}
        />
      </div>
    );

  return isMobile ? renderSides() : <div>{renderSides()}</div>;
}

export default PolygonOptions;
