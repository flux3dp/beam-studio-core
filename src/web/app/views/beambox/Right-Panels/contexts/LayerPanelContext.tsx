import * as React from 'react';
import { ILayerPanelContext } from 'interfaces/IContext';

export const LayerPanelContext = React.createContext<ILayerPanelContext>(null);

interface Props {
  children: React.ReactNode;
}

export class LayerPanelContextProvider extends React.Component<Props> {
  private contextValue = null;

  private selectedLayers: string[] = [];

  constructor(props) {
    super(props);
    this.updateContextValue();
  }

  updateContextValue = (): void => {
    const {
      selectedLayers,
      setSelectedLayers,
      updateLayerPanel,
    } = this;
    this.contextValue = {
      selectedLayers,
      setSelectedLayers,
      updateLayerPanel,
    };
  };

  setSelectedLayers = (selectedLayers: string[]): null => {
    this.selectedLayers = [...selectedLayers];
    this.updateContextValue();
    this.forceUpdate();
    return null;
  };

  updateLayerPanel = (): void => {
    this.forceUpdate();
  };

  render() {
    const { children } = this.props;
    return (
      <LayerPanelContext.Provider value={this.contextValue}>
        {children}
      </LayerPanelContext.Provider>
    );
  }
}
