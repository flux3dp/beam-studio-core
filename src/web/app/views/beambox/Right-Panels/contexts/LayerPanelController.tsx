import eventEmitterFactory from 'helpers/eventEmitterFactory';

const layerPanelEventEmitter = eventEmitterFactory.createEventEmitter('layer-panel');

const updateLayerPanel = (): void => {
  layerPanelEventEmitter.emit('UPDATE_LAYER_PANEL');
};

const getSelectedLayers = (): string[] => {
  const response = {
    selectedLayers: [],
  };
  layerPanelEventEmitter.emit('GET_SELECTED_LAYERS', response);
  return response.selectedLayers;
};

const setSelectedLayers = (selectedLayers: string[]): void => {
  layerPanelEventEmitter.emit('SET_SELECTED_LAYERS', selectedLayers);
};

export default {
  updateLayerPanel,
  getSelectedLayers,
  setSelectedLayers,
};
