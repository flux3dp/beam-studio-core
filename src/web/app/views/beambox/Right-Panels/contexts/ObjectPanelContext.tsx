import React from 'react';

import eventEmitterFactory from 'helpers/eventEmitterFactory';

interface IObjectPanelContext {
  polygonSides: number;
  dimensionValues: any;
  updateDimensionValues: (newValues: any) => void;
  getDimensionValues: (response: {
    dimensionValues: any,
  }, key?: string,) => any;
  updateObjectPanel: () => void;
}

export const ObjectPanelContext = React.createContext<IObjectPanelContext>({
  polygonSides: 5,
  dimensionValues: {},
  updateDimensionValues: () => { },
  getDimensionValues: () => { },
  updateObjectPanel: () => { },
});
const objectPanelEventEmitter = eventEmitterFactory.createEventEmitter('object-panel');
const minRenderInterval = 200;

interface State {
  lastUpdateTime: number;
  polygonSides: number;
}

export class ObjectPanelContextProvider extends React.Component<any, State> {
  private dimensionValues: any;

  private updateTimeout: NodeJS.Timeout;

  constructor(props) {
    super(props);
    this.dimensionValues = {};
    this.state = {
      lastUpdateTime: Date.now(),
      polygonSides: 5,
    };
  }

  componentDidMount() {
    objectPanelEventEmitter.on('UPDATE_DIMENSION_VALUES', this.updateDimensionValues.bind(this));
    objectPanelEventEmitter.on('GET_DIMENSION_VALUES', this.getDimensionValues.bind(this));
    objectPanelEventEmitter.on('UPDATE_OBJECT_PANEL', this.updateObjectPanel.bind(this));
    objectPanelEventEmitter.on('UPDATE_POLYGON_SIDES', this.updatePolygonSides.bind(this));
  }

  componentWillUnmount() {
    objectPanelEventEmitter.removeAllListeners();
  }

  updateDimensionValues = (newValues: any): void => {
    this.dimensionValues = {
      ...this.dimensionValues,
      ...newValues,
    };
  };

  updatePolygonSides = (polygonSides: number): void => {
    this.setState({
      polygonSides,
    });
  };

  getDimensionValues = (response: {
    dimensionValues: any,
  }, key?: string): void => {
    response.dimensionValues = key ? this.dimensionValues[key] : this.dimensionValues;
  };

  updateObjectPanel = (): void => {
    clearTimeout(this.updateTimeout);
    const time = Date.now();
    const { lastUpdateTime } = this.state;
    if (time - lastUpdateTime >= minRenderInterval) {
      this.setState({ lastUpdateTime: time });
    } else {
      this.updateTimeout = setTimeout(() => {
        this.setState({
          lastUpdateTime: lastUpdateTime + minRenderInterval,
        });
      }, lastUpdateTime + minRenderInterval - time);
    }
  };

  render(): JSX.Element {
    const { children } = this.props;
    const { polygonSides } = this.state;
    const {
      dimensionValues,
      updateDimensionValues,
      getDimensionValues,
      updateObjectPanel,
    } = this;
    return (
      <ObjectPanelContext.Provider value={{
        polygonSides,
        dimensionValues,
        updateDimensionValues,
        getDimensionValues,
        updateObjectPanel,
      }}
      >
        {children}
      </ObjectPanelContext.Provider>
    );
  }
}
