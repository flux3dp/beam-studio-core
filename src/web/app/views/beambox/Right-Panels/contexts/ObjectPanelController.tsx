import eventEmitterFactory from 'helpers/eventEmitterFactory';

const objectPanelEventEmitter = eventEmitterFactory.createEventEmitter('object-panel');

const updateDimensionValues = (newValue: any): void => {
  objectPanelEventEmitter.emit('UPDATE_DIMENSION_VALUES', newValue);
};

const getDimensionValues = (key: string): any => {
  const response = {
    dimensionValues: {},
  };
  objectPanelEventEmitter.emit('GET_DIMENSION_VALUES', key, response);
  return response.dimensionValues;
};

const updateObjectPanel = (): void => {
  objectPanelEventEmitter.emit('UPDATE_OBJECT_PANEL');
};

export default {
  updateObjectPanel,
  updateDimensionValues,
  getDimensionValues,
};
