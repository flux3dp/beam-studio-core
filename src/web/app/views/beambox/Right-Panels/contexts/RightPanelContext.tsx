import * as React from 'react';

export const RightPanelContext = React.createContext(null);

interface State {
  mode: string;
  selectedElement: Element;
}

export class RightPanelContextProvider extends React.Component<any, State> {
    constructor(props) {
        super(props);
        this.state = {
            mode: 'element',
            selectedElement: null,
        }
    }

    setMode = (mode) => {
        const {mode: currentMode} = this.state;
        if (['path-edit'].includes(mode) || currentMode !== mode) {
            this.setState({ mode });
        }
    }

    setSelectedElement = (elems) => {
        if (elems !== this.state.selectedElement) {
            (document.activeElement as HTMLInputElement).blur();
        }
        this.setState({selectedElement: elems});
    }

    render() {
        const {
            setMode,
            setSelectedElement,
        } = this;
        const {
            mode,
            selectedElement,
        } = this.state;
        return (
            <RightPanelContext.Provider value={{
                setMode,
                mode,
                setSelectedElement,
                selectedElement,
            }}>
                {this.props.children}
            </RightPanelContext.Provider>
        );
    }
};
