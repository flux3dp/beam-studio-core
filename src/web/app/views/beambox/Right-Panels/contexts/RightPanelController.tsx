import BeamboxGlobalInteraction from 'app/actions/beambox/beambox-global-interaction';
import { RightPanelContextHelper } from 'app/views/beambox/Right-Panels/Right-Panel';

const setSelectedElement = (elem) => {
    if (!elem) {
        BeamboxGlobalInteraction.onObjectBlur();
    } else {
        BeamboxGlobalInteraction.onObjectBlur();
        BeamboxGlobalInteraction.onObjectFocus([elem]);
    }
    if (!RightPanelContextHelper.context) {
        console.log('RightPanel is not mounted now.');
    } else {
        RightPanelContextHelper.context.setSelectedElement(elem);
    }
};

const toPathEditMode = () => {
    if (!RightPanelContextHelper.context) {
        console.log('RightPanel is not mounted now.');
    } else {
        RightPanelContextHelper.context.setMode('path-edit');
    }
}

const toElementMode = () => {
    if (!RightPanelContextHelper.context) {
        console.log('RightPanel is not mounted now.');
    } else {
        RightPanelContextHelper.context.setMode('element');
    }
}

export default {
    setSelectedElement,
    toPathEditMode,
    toElementMode,
}
