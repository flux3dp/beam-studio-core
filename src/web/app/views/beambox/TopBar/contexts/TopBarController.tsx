import eventEmitterFactory from 'helpers/eventEmitterFactory';

const topBarEventEmitter = eventEmitterFactory.createEventEmitter('top-bar');

const updateTopBar = (): void => {
  topBarEventEmitter.emit('UPDATE_TOP_BAR');
};

const setElement = (elem: Element | null): void => {
  topBarEventEmitter.emit('SET_ELEMENT', elem);
};

const setFileName = (fileName: string): void => {
  topBarEventEmitter.emit('SET_FILE_NAME', fileName);
};

const setHasUnsavedChange = (hasUnsavedChange: boolean): void => {
  topBarEventEmitter.emit('SET_HAS_UNSAVED_CHANGE', hasUnsavedChange);
};

const getTopBarPreviewMode = (): boolean => {
  const response = {
    isPreviewMode: false,
  };
  topBarEventEmitter.emit('GET_TOP_BAR_PREVIEW_MODE', response);
  return response.isPreviewMode;
};

const setShouldStartPreviewController = (shouldStartPreviewController: boolean): void => {
  topBarEventEmitter.emit('SET_SHOULD_START_PREVIEW_CONTROLLER', shouldStartPreviewController);
};

const setStartPreviewCallback = (callback: () => void): void => {
  topBarEventEmitter.emit('SET_START_PREVIEW_CALLBACK', callback);
};

export default {
  updateTopBar,
  setElement,
  setFileName,
  setHasUnsavedChange,
  getTopBarPreviewMode,
  setShouldStartPreviewController,
  setStartPreviewCallback,
};
