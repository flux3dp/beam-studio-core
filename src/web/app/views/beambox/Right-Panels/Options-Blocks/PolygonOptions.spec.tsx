import React from 'react';
import { fireEvent, render, waitFor } from '@testing-library/react';

import { ObjectPanelContextProvider } from 'app/views/beambox/Right-Panels/contexts/ObjectPanelContext';

import PolygonOptions from './PolygonOptions';

jest.mock('app/views/beambox/Right-Panels/Options-Blocks/InFillBlock', () => () => <div>DummyInFillBlock</div>);

const useIsMobile = jest.fn();
jest.mock('helpers/system-helper', () => ({
  useIsMobile: () => useIsMobile(),
}));

jest.mock('helpers/i18n', () => ({
  lang: {
    beambox: {
      right_panel: {
        object_panel: {
          option_panel: {
            sides: 'Sides',
          },
        },
      },
    },
  },
}));

window.polygonAddSides = jest.fn(() => {
  const sides = +document.getElementById('flux').getAttribute('sides');
  document.getElementById('flux').setAttribute('sides', (sides + 1).toString());
});
window.polygonDecreaseSides = jest.fn(() => {
  const sides = +document.getElementById('flux').getAttribute('sides');
  document.getElementById('flux').setAttribute('sides', Math.max((sides - 1),3).toString());
});

describe('test PolygonOptions', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  test('should render correctly', () => {
    const { container, rerender } = render(
      <PolygonOptions
        elem={document.getElementById('flux')}
        polygonSides={0}
      />
    );
    expect(container).toMatchSnapshot();

    const elem = document.createElement('polygon');
    elem.setAttribute('id', 'flux');
    elem.setAttribute('sides', '5');
    document.body.appendChild(elem);
    rerender(
      <PolygonOptions
        elem={document.getElementById('flux')}
        polygonSides={5}
      />
    );

    expect(container).toMatchSnapshot();

    const input = container.querySelector('input');
    fireEvent.change(input, { target: { value: 8 } });
    fireEvent.blur(input);

    expect(container).toMatchSnapshot();
    expect(window.polygonAddSides).toHaveBeenCalledTimes(3);
    expect(window.polygonDecreaseSides).not.toHaveBeenCalled();

    jest.clearAllMocks();

    fireEvent.change(input, { target: { value: 5 } });
    fireEvent.blur(input);
    expect(container).toMatchSnapshot();
    expect(window.polygonAddSides).not.toHaveBeenCalled();
    expect(window.polygonDecreaseSides).toHaveBeenCalledTimes(3);

    jest.clearAllMocks();

    fireEvent.change(input, { target: { value: 5 } });
    fireEvent.blur(input);
    expect(window.polygonAddSides).not.toHaveBeenCalled();
    expect(window.polygonDecreaseSides).not.toHaveBeenCalled();
  });

  test('should render correctly in mobile', async() => {
    useIsMobile.mockReturnValue(true)
    const { baseElement, container, getByText, rerender } = render(
      <PolygonOptions
        elem={document.getElementById('flux')}
        polygonSides={0}
      />
    );
    expect(container).toMatchSnapshot();

    const elem = document.createElement('polygon');
    elem.setAttribute('id', 'flux');
    elem.setAttribute('sides', '5');
    document.body.appendChild(elem);
    rerender(
      <ObjectPanelContextProvider>
        <PolygonOptions
          elem={document.getElementById('flux')}
          polygonSides={5}
        />
      </ObjectPanelContextProvider>
    );
    expect(container).toMatchSnapshot();

    const objectPanelItem = baseElement.querySelector('div.object-panel-item');
    const displayBtn = baseElement.querySelector('button.number-item');
    expect(displayBtn).toHaveTextContent('5');
    expect(objectPanelItem).not.toHaveClass('active');
    fireEvent.click(objectPanelItem);
    await waitFor(() => expect(baseElement.querySelector('.adm-mask')).toBeVisible());
    expect(objectPanelItem).toHaveClass('active');
    expect(getByText('.').parentElement).toHaveClass('adm-button-disabled');
    expect(window.polygonAddSides).not.toHaveBeenCalled();
    expect(window.polygonDecreaseSides).not.toHaveBeenCalled();

    fireEvent.click(baseElement.querySelectorAll('.input-keys button')[11]);
    expect(window.polygonAddSides).not.toHaveBeenCalled();
    expect(window.polygonDecreaseSides).toHaveBeenCalledTimes(5);
    expect(expect(displayBtn).toHaveTextContent('3'));
  });
})
