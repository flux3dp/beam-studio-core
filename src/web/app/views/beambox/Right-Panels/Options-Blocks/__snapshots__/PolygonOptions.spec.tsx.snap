// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<div
  className="polygon-options"
>
  <div
    className="option-block"
    key="polygon-sides"
  >
    <div
      className="label"
    >
      Sides
    </div>
    <UnitInput
      abbr={false}
      className={
        {
          "option-input": true,
        }
      }
      decimal={0}
      defaultValue={5}
      disabled={false}
      displayMultiValue={false}
      forceUsePropsUnit={false}
      getValue={[Function]}
      isDoOnInput={false}
      max={9007199254740991}
      min={3}
      onBlur={[Function]}
      onFocus={[Function]}
      onKeyUp={[Function]}
      step={1}
      type="text"
      unit=""
    />
  </div>
  <InFillBlock
    elem={null}
  />
</div>
`;

exports[`should render correctly 2`] = `
<div
  className="polygon-options"
>
  <div
    className="option-block"
    key="polygon-sides"
  >
    <div
      className="label"
    >
      Sides
    </div>
    <UnitInput
      abbr={false}
      className={
        {
          "option-input": true,
        }
      }
      decimal={0}
      defaultValue={5}
      disabled={false}
      displayMultiValue={false}
      forceUsePropsUnit={false}
      getValue={[Function]}
      isDoOnInput={false}
      max={9007199254740991}
      min={3}
      onBlur={[Function]}
      onFocus={[Function]}
      onKeyUp={[Function]}
      step={1}
      type="text"
      unit=""
    />
  </div>
  <InFillBlock
    elem={
      <div
        id="flux"
      />
    }
  />
</div>
`;

exports[`should render correctly 3`] = `
<div
  className="polygon-options"
>
  <div
    className="option-block"
    key="polygon-sides"
  >
    <div
      className="label"
    >
      Sides
    </div>
    <UnitInput
      abbr={false}
      className={
        {
          "option-input": true,
        }
      }
      decimal={0}
      defaultValue={8}
      disabled={false}
      displayMultiValue={false}
      forceUsePropsUnit={false}
      getValue={[Function]}
      isDoOnInput={false}
      max={9007199254740991}
      min={3}
      onBlur={[Function]}
      onFocus={[Function]}
      onKeyUp={[Function]}
      step={1}
      type="text"
      unit=""
    />
  </div>
  <InFillBlock
    elem={
      <div
        id="flux"
      />
    }
  />
</div>
`;

exports[`should render correctly 4`] = `
<div
  className="polygon-options"
>
  <div
    className="option-block"
    key="polygon-sides"
  >
    <div
      className="label"
    >
      Sides
    </div>
    <UnitInput
      abbr={false}
      className={
        {
          "option-input": true,
        }
      }
      decimal={0}
      defaultValue={5}
      disabled={false}
      displayMultiValue={false}
      forceUsePropsUnit={false}
      getValue={[Function]}
      isDoOnInput={false}
      max={9007199254740991}
      min={3}
      onBlur={[Function]}
      onFocus={[Function]}
      onKeyUp={[Function]}
      step={1}
      type="text"
      unit=""
    />
  </div>
  <InFillBlock
    elem={
      <div
        id="flux"
      />
    }
  />
</div>
`;
